packages/vuetify/src/components/VBadge/VBadge.ts

```diff
@@ -1,72 +1,202 @@
 // Styles
 import './VBadge.sass'
 
+// Components
+import VIcon from '../VIcon/VIcon'
+
 // Mixins
 import Colorable from '../../mixins/colorable'
+import Themeable from '../../mixins/themeable'
 import Toggleable from '../../mixins/toggleable'
-import { factory as PositionableFactory } from '../../mixins/positionable'
 import Transitionable from '../../mixins/transitionable'
+import { factory as PositionableFactory } from '../../mixins/positionable'
+
+// Utilities
+import mixins from '../../util/mixins'
+import {
+  convertToUnit,
+  getSlot,
+  getObjectValueByPath,
+} from '../../util/helpers'
 
 // Types
 import { VNode } from 'vue'
-import mixins from '../../util/mixins'
 
 export default mixins(
   Colorable,
-  Toggleable,
   PositionableFactory(['left', 'bottom']),
-  Transitionable
+  Themeable,
+  Toggleable,
+  Transitionable,
 /* @vue/component */
 ).extend({
   name: 'v-badge',
 
   props: {
+    avatar: Boolean,
+    bordered: Boolean,
     color: {
       type: String,
       default: 'primary',
     },
+    content: { required: false },
+    dot: Boolean,
+    label: {
+      type: String,
+      default: '$vuetify.badge',
+    },
+    icon: String,
+    inline: Boolean,
+    offsetX: [Number, String],
+    offsetY: [Number, String],
     overlap: Boolean,
+    tile: Boolean,
     transition: {
       type: String,
-      default: 'fab-transition',
-    },
-    value: {
-      default: true,
+      default: 'scale-rotate-transition',
     },
+    value: { default: true },
   },
 
   computed: {
     classes (): object {
       return {
+        'v-badge--avatar': this.avatar,
+        'v-badge--bordered': this.bordered,
         'v-badge--bottom': this.bottom,
+        'v-badge--dot': this.dot,
+        'v-badge--icon': this.icon != null,
+        'v-badge--inline': this.inline,
         'v-badge--left': this.left,
         'v-badge--overlap': this.overlap,
+        'v-badge--tile': this.tile,
+        ...this.themeClasses,
+      }
+    },
+    computedBottom (): string {
+      return this.bottom ? 'auto' : this.computedYOffset
+    },
+    computedLeft (): string {
+      if (this.isRtl) {
+        return !this.left ? 'auto' : this.computedXOffset
+      }
+
+      return this.left ? 'auto' : this.computedXOffset
+    },
+    computedRight (): string {
+      if (this.isRtl) {
+        return this.left ? 'auto' : this.computedXOffset
+      }
+
+      return !this.left ? 'auto' : this.computedXOffset
+    },
+    computedTop (): string {
+      return !this.bottom ? 'auto' : this.computedYOffset
+    },
+    computedXOffset (): string {
+      return this.calcPosition(this.offsetX)
+    },
+    computedYOffset (): string {
+      return this.calcPosition(this.offsetY)
+    },
+    isRtl (): boolean {
+      return getObjectValueByPath(this.$vuetify, 'rtl', false)
+    },
+    // Default fallback if offsetX
+    // or offsetY are undefined.
+    offset (): number {
+      if (this.overlap) return this.dot ? 8 : 12
```

Comment:
I wonder if there's a way (probably not trivial, but imho should be possible - the question would be is it worth it)  to put it in .scss (and if yes - add variables)

Reply:
Possible but not part of this PR. Let's do that when we start working on the Grid having linked js/sass logic.